### PR TITLE
Update to `V4` 🚀 

### PR DESCRIPTION
**Pull request by terencechain**
⚠️ Please do not merge this until we have 2-3 approvals ⚠️ 

Steps taken:

1.)
In Goland, search for all files with file mask `*.bazel` and change `github.com/prysmaticlabs/prysm/v3` to `github.com/prysmaticlabs/prysm/v4 `

2.)
Do the same for `*.go` and `*.proto` files

3.)
Do the same for `update-mockgen.sh`

4.)
Do the same for `hack/update-go-pbs.sh`

5.)
Update `go.mod` `module module github.com/prysmaticlabs/prysm/v3` to `v4`

6.)
Update `.deepsource.toml` `import_paths = ["github.com/prysmaticlabs/prysm/v3"]` to `v4`

7.)
Our grpc gateway, which is a fork we maintain here was hardcoding v3 paths in a template generator. It was updated in a commit here: https://github.com/prysmaticlabs/grpc-gateway/pull/11

Update `go.mod` and `go.sum`

`go get  github.com/prysmaticlabs/grpc-gateway/v2@09284ba20446ebc03dd4cbd5a0c860259b8e5984`

Use `v2.3.1-0.20230315201114-09284ba20446` in `go.mod`

8.)
Run gazella
`bazel run //:gazelle -- update-repos -from_file=go.mod -to_macro=deps.bzl%prysm_deps -prune=true`

9.)
Run `go mod tidy`

10.)
Run script `go get ./... ` 

11.)
Run script`./hack/update-go-pbs.sh`

12.)
Run script `./hack/update-go-ssz.sh` 

13.)
Pass `go build ./...`

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Did you run `./hack/update-go-ssz.sh`?

**Comment by terencechain** [comment]
> Did you run `./hack/update-go-ssz.sh`?

Missed this. I ran it and updated the steps. Thank you

**Comment by prestonvanloon** [comment]
Great work! Looks like some build failures though.

**Review by james-prysm — APPROVED** [review]
[james-prysm-v4](https://github.com/prysmaticlabs/prysm/compare/develop...james-prysm-v4) got similar results with blobs generated file as the extra file.

File: hack/update-mockgen.sh [hunk 1]
```diff
@@ -35,7 +35,7 @@ for ((i = 0; i < ${#proto_mocks_service[@]}; i++)); do
     interfaces=${proto_mocks_service[i]#* };
     echo "generating $file for interfaces: $interfaces";
     echo
-    GO11MODULE=on mockgen -package=mock -destination="$file" github.com/prysmaticlabs/prysm/v3/proto/eth/service "$interfaces"
+    GO11MODULE=on mockgen -package=mock -destination="$file" github.com/prysmaticlabs/prysm/v4/proto/eth/service "$interfaces"
 done
 
 # github.com/prysmaticlabs/prysm/proto/v3/prysm/v1alpha1/validator-client
```
Comment (prestonvanloon) [inline hack/update-mockgen.sh:41]: Want to fix this too?

Comment (terencechain) [reply hack/update-mockgen.sh:41]: Ah. Nice catch

**Review by potuz — APPROVED** [review]
I ran different changes with sed magic since I don't have goland. I get a different branch than Terence, but the difference seems to be innocuous. Check against the branch `potuz-v4-vis` please.

**Review by prestonvanloon — APPROVED** [review]
LGTM. I checked @potuz's branch diff too. looks fine, some difference in deps.bzl ordering and potuz's branch needs to run goimports or manually fix imports that were arranged incorrectly for pb.go.

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**